### PR TITLE
Feature/core utils

### PR DESCRIPTION
**Pull request by AlexandrZagorskiy**

### CODE REVIEW

File: src/sc_kpm/utils/search_utils.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def get_first_search_template_result(template: ScTemplate) -> Optional[ScTemplateResult]:
+    search_results = client.template_search(template)
+    if len(search_results) > 0:
+        return search_results[0]
+    return None
+
+
+def get_content_from_links_set(links_set_node: ScAddr) -> str:
```
Comment (NikitaZotov) [inline src/sc_kpm/utils/search_utils.py:71]: Rename this method. It is not meaningful

File: tests/common_tests.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+(See an accompanying file LICENSE or a copy at https://opensource.org/licenses/MIT)
+"""
+from unittest import TestCase
+
+import sc_client.client
+
+
+class BaseTestCase(TestCase):
+    def setUp(self) -> None:
+        sc_client.client.connect("ws://localhost:8090/ws_json")
```
Comment (NikitaZotov) [inline tests/common_tests.py:13]: make constant

File: tests/test_utils/test_common_utils.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    get_edge,
+    get_edges,
+    get_element_by_role_relation,
+    get_link_content,
+    get_system_idtf,
+)
+from tests.common_tests import BaseTestCase
+
+
+class TestActionUtils(BaseTestCase):
```
Comment (NikitaZotov) [inline tests/test_utils/test_common_utils.py:31]: Add another tests

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+name: Tests
+
+on: [push, pull_request]
```
[the remark this answers is not in the document]
Comment (AlexandrZagorskiy) [reply .github/workflows/tests.yml:3]: Left _pull_request_ here

File: src/sc_kpm/utils/search_utils.py (hunk 1 above)
Comment (AlexandrZagorskiy) [reply src/sc_kpm/utils/search_utils.py:71]: removed it

File: src/sc_kpm/common/sc_agent.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class ScAgent(ScAgentAbstract):
+    @classmethod
+    @abstractmethod
+    def setup(cls):
+        raise NotImplementedError
+
+    @abstractmethod
+    def on_event(self, source_node: ScAddr, edge: ScAddr, target_node: ScAddr) -> None:
```
Comment (NikitaZotov) [inline src/sc_kpm/common/sc_agent.py:63]:
```suggestion
    def on_event(self, source_node: ScAddr, edge: ScAddr, target_node: ScAddr) -> ScResult:
```

File: src/sc_kpm/common/sc_module.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                raise RuntimeError("Cannot register agents: connection to the sc-server is not established")
+
+    def unregister(self) -> None:
+        for agent in self.agents:
+            agent.unregister()
+        self.is_registred = False
+        print(f"{self.__class__.__name__} is unregistred")
+
+
+class ScServer:
```
Comment (NikitaZotov) [inline src/sc_kpm/common/sc_module.py:42]: Make the package for this class

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: src/sc_kpm/common/sc_module.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+from sc_client import client
+
+from sc_kpm.common.sc_agent import ScAgent
+
+
+class ScModule(ABC):
+    agents: List[ScAgent] = []
+
+    def __init__(self):
+        self.is_registred = False
```
Comment (FalseR20) [inline src/sc_kpm/common/sc_module.py:21]:
```suggestion
        self.is_registered = False
```

File: README.md [hunk 8]
```diff
@@ -1,2 +1,95 @@
...
+```
+
+After stopping server, all modules and agents will be deactivated.
+```
+server.stop()
+...
+```
+
+## Utils
+...
```
Comment (NikitaZotov) [inline README.md:95]: Add utils descriptions

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

**Comment by NikitaZotov** [comment]
Add issue about sc-server usage in tests

**Review by NikitaZotov — APPROVED** [review]

**Pull request merged**